WizardAskPassword: add colors to password strength meter; add "Passwords match/do not match" message

rating89us opened the pull request:
Changes:
- Move password strength meter below the password field
- Password strength meter bar changes color from red (low) -> yellow (medium) -> green (high)
- "Password (confirm)" field borders change to red when password doesn't match
- Add "password match/doesn't match" message below the confirmation password field

Success example: the user types a strong password (password strength bar changes color from red -> yellow -> green). Next, the user types a password confirmation that matches. No error is displayed, everything is green, and "Next" button is enabled.
![success](https://user-images.githubusercontent.com/45968869/131997046-58f7f7b7-86fa-42a9-8af5-b40e48180226.gif)

Failed example: the user types a partial password confirmation that doesn't match. When the user types a wrong character, "Password (confirm)" field shows red borders and a red error message is displayed. "Next" button stays disabled. The backspace key can be used to correct the typing error.
![fail](https://user-images.githubusercontent.com/45968869/131997053-fa345380-7bcc-4ed7-886c-1ab34a5f9012.gif)

selsta commented:
<img width="1026" alt="Screenshot 2021-09-03 at 00 14 27" src="https://user-images.githubusercontent.com/7697454/131923166-902fed34-e181-4fe5-8450-b262818a56fa.png">

This doesn't look good, everything on the Wizard page is full width so having this not stands out.

Especially if you consider that now a 32 character length password doesn't fit in.

<img width="523" alt="Screenshot 2021-09-03 at 00 17 30" src="https://user-images.githubusercontent.com/7697454/131923424-9eae35ba-94ba-4cff-a8fa-5490c5e73431.png">

We have to space so no reason to make it artificially small.

Will review code separately.

selsta reviewed (commented):
```diff
diff --git a/wizard/WizardAskPassword.qml b/wizard/WizardAskPassword.qml
index ed88c8cb..023d0df9 100644
--- a/wizard/WizardAskPassword.qml
+++ b/wizard/WizardAskPassword.qml
@@ -40,45 +40,9 @@ ColumnLayout {
     property alias password: passwordInput.text
     property int passwordFill: 0
     property string passwordStrengthText: qsTr("Strength: ") + translationManager.emptyString
-    property bool passwordDoesNotMatch: false
 
     function calcStrengthAndVerify(){
         calcPasswordStrength();
-        if (passwordInputConfirm.text != "") {
-            if (passwordInput.text.substring(0, passwordInputConfirm.text.length) != passwordInputConfirm.text) {
-                errorMessagePasswordInputConfirm.text = qsTr("Password does not match") + translationManager.emptyString;
-                root.passwordDoesNotMatch = true;
-            } else {
-                if (passwordInputConfirm.text.length != passwordInput.text.length) {
-                    if (passwordInputConfirm.firstUserInput) {
-                        errorMessagePasswordInputConfirm.text = "";
-                        root.passwordDoesNotMatch = true;
-                    } else {
-                        errorMessagePasswordInputConfirm.text = qsTr("Password does not match") + translationManager.emptyString;
-                        root.passwordDoesNotMatch = true;
-                    }
-                } else {
-                    errorMessagePasswordInputConfirm.text = qsTr("Password match!") + translationManager.emptyString;
-                    root.passwordDoesNotMatch = false;
-                }
-            }
-        } else {
-            if (passwordInput.text === passwordInputConfirm.text) {
-                if (passwordInputConfirm.firstUserInput) {
-                    root.passwordDoesNotMatch = false;
-                } else {
-                    errorMessagePasswordInputConfirm.text = "";
-                    root.passwordDoesNotMatch = false;
-                }
-            } else {
-                if (passwordInputConfirm.firstUserInput) {
-                    root.passwordDoesNotMatch = true;
-                } else {
-                    errorMessagePasswordInputConfirm.text = qsTr("Password does not match") + translationManager.emptyString;
-                    root.passwordDoesNotMatch = true;
-                }
-            }
-        }
         return passwordInput.text === passwordInputConfirm.text;
     }
 
@@ -205,8 +169,8 @@ ColumnLayout {
             Layout.fillWidth: true
             Layout.topMargin: 8
             KeyNavigation.tab: passwordInputConfirm
-            error: root.passwordDoesNotMatch && errorMessagePasswordInputConfirm.text != ""
-            errorWhenEmpty: root.passwordDoesNotMatch && errorMessagePasswordInputConfirm.text != ""
+            error: passwordInputMessage.passwordMatch && passwordInputMessage.visible
+            errorWhenEmpty: passwordInputMessage.passwordMatch && passwordInputMessage.visible
             labelFontSize: 14
             passwordLinked: passwordInput
             labelText: qsTr("Password (confirm)") + translationManager.emptyString
@@ -224,22 +188,23 @@ ColumnLayout {
             Layout.maximumWidth: 150
 
             MoneroComponents.TextPlain {
-                id: iconErrorMessagePasswordInputConfirm
-                visible: errorMessagePasswordInputConfirm.text != ""
+                visible: passwordInputMessage.visible
                 font.family: FontAwesome.fontFamilySolid
                 font.styleName: "Solid"
                 font.pixelSize: 15
-                text: root.passwordDoesNotMatch ? FontAwesome.exclamationCircle : FontAwesome.checkCircle
-                color: errorMessagePasswordInputConfirm.color
+                text: !passwordInputMessage.passwordMatch ? FontAwesome.exclamationCircle : FontAwesome.checkCircle
+                color: passwordInputMessage.color
                 themeTransition: false
             }
 
             MoneroComponents.TextPlain {
-                id: errorMessagePasswordInputConfirm
+                id: passwordInputMessage
+                property bool passwordMatch: passwordInputConfirm.text === passwordInput.text
+                visible: passwordInputConfirm.text.length > 0 && !passwordInputConfirm.firstUserInput
                 Layout.topMargin: 3
-                text: "";
+                text: passwordMatch ? qsTr("Password match!") : qsTr("Password does not match");
                 textFormat: Text.PlainText
-                color: root.passwordDoesNotMatch ? "#FF0000" : (MoneroComponents.Style.blackTheme ? "#00FF00" : "#008000")
+                color: !passwordMatch ? "#FF0000" : (MoneroComponents.Style.blackTheme ? "#00FF00" : "#008000")
                 font.family: MoneroComponents.Style.fontRegular.name
                 font.pixelSize: 14
                 themeTransition: false
```

This should be the same as your code, just simplified.

rating89us commented:
Thanks for reviewing.

I changed your code to verify for partial password (on first user input). Thinking in Android/mobile UI/UX, I moved password strength bar below password field and error/success message below the confirmation password field. Also made there items fill screen width.

selsta commented:
Looks good now :) Please squash, then we can merge it.
